### DISCUSSION TITLE
How to take historical data into account in loadBalance collector.

### DISCUSSION THREAD

**Discussion by Boyankovachev** [TimefoldAI/timefold-solver#1127] [General]
Hello. I am trying to implement a fairness constraint for employees working night shifts. My constraint works great and is very similar to the one in the example in the docs [here](https://docs.timefold.ai/timefold-solver/latest/constraints-and-score/load-balancing-and-fairness#fairnessConstraints).
My problem is that I am generating a plan for a given month, but I need take into account historical data for the given employee, meaning I have a getter in the Employee, getPastNightShifts(), which returns me the number of night shifts he has worked in the previous X months. I would like to pass the data somehow into the loadBalance constraint collector, so they are also taken into consideration (added to the number of night shift in the month being generated). 
Is there currently any way this can be achieved through the Timefold API? If not, I think it would be a nice new feature, and got any tips how I can implement it myself? Thank you!

```
    public Constraint justiceByNightShifts(ConstraintFactory factory){
        return factory.forEach(ShiftAssignment.class)
                .filter((shiftAssignment -> shiftAssignment.getShift().isNightShift()))
                .groupBy(
                        ConstraintCollectors.loadBalance(ShiftAssignment::getEmployee)
                )
                .penalizeBigDecimal(ConstraintNames.JUSTICE_BY_NIGHT_SHIFTS.getDefaultValue(), LoadBalance::unfairness)
                .asConstraint(ConstraintNames.JUSTICE_BY_NIGHT_SHIFTS.getName());
    }
```

**Comment by triceo** [comment]
Hello @Boyankovachev - yes, there is.
`ConstraintCollectors.loadBalance(...)` has an overload which (among other arguments) also accepts the initial load - which exists precisely to solve your problem.

**Reply by Boyankovachev** [reply]
Perfect, thank you!

**Comment by ge0ffrey** [comment]
More generally, this is the "tally" design pattern, which the old tally (how many xmas's Ann had to work in the past 10 years, or weekends in the last 2 years) gets added in the dataset without the actual shifts of so long ago.
